### ISSUE TITLE
Spells should not strip characters like * or (self only)

### ISSUE THREAD

**Issue by mlenser** [symposion/roll20-shaped-scripts#261]
These cases are important to show to the user and shouldn't be stripped.
See https://bitbucket.org/mlenser/5eshaped/issues/71/monster-spell-import-issues

I forget if this is a script issue or a json data issue. If the latter I think we may need to adjust how that data is formed.

**Comment by mlenser** [comment]
Nice to have

**Comment by mlenser** [comment]
https://lh3.googleusercontent.com/-bkEP82GTVIg/WHHvzbhc-nI/AAAAAAAAYj4/HzNGVQwKPKsII1qrsNwjdV0Xoo6BACGhQCK8B/s512/2017-01-07.png for example shows the problem. Spiritual Weapon has other characters so my check fails. Spiritual Weapon should come in with the "(Spear)" part.

**Comment by symposion** [comment]
I'll look into this but it's likely to be fiddly. The problem is that the spell is "Spiritual Weapon" or "Mage Armor", and this is how it needs to be looked up. The thing in brackets is a restriction to the usage of the spell by this particular user, rather than a distinct spell.

I strongly disagree with the idea that the * should be incorporated into the spell name - this is clearly wrong to me, since this has nothing to do with the spell itself, but is merely a note to say that this spell is cast before combat. 

For me the correct solution to this problem is to have an annotations field of some sort in the sheet that allows some sort of notes to be provided that qualify the spells usage by this character if appropriate, rather than polluting the name field. The original bug shows the problem - if you start putting e.g. the asterisk in the name then it becomes difficult to identify programmatically which spells a character already has because you can't just search for the spell name, you have to account for the possibility that it might have any number of other random text at the end as well.

It's also worth noting that this information will remain visible in the spellcasting trait for NPCs, even if it's stripped out in the spell list. I'd argue that this is actually the right place for it.

**Comment by mlenser** [comment]
The spellcasting trait is now parsed completely and removed. Spells only remain for innate spellcasting.

Currently we're stripping out vital information. Whether that's (self-only) or the star or the (Trident). That should not be the case. Parsing that data to another field could be an option to consider, but it definitely complicates some things. I can handle it a bit like recharge for traits if we decide to go that route.

**Comment by symposion** [comment]
Ok, if the spellcasting trait is now stripped out I agree we definitely need another solution - this information is important and needs to be retained somewhere. I can see a case for (self-only) or (Trident) being part of the spell name, since these are limitations on the function of the spell and make it different to the unlimited version. The asterisk is merely a note about how this character typically uses this spell and I really don't think it belongs in the name of the spell.

Either way, retaining this is in the spell name is going to be a bit ugly for me to implement since I look up the spells by name in a map using generic code; I'll have to reimplement this to strip certain special characters, find the spell, and then re-add the characters to the loaded spell name afterwards. Because the code that does is a generic thing that also handles loading monsters from the JSON database it's going to be messy and take a little while to get right.

On the other hand, I guess that the alternative is probably worse - which is that you'd have to parse the information out of the spellcasting trait, locate the script-imported spells that correspond and then update them.

In the short term there's a possibility that I can achieve this using the same functionality as ♯255 - having the option to specify and alias when importing. I'll look into that but I'm still uncomfortable about the asterisk thing - it feels wrong having that as part of the spell name. If you can come up with a better solution for that in the sheet for the future that would be good.

**Comment by mlenser** [comment]
I suggest we wait on this until you have some more time. We can discuss all the possible options. Either way it'll create work for either you or me and either is fine by me - I just want the best solution.

**Comment by symposion** [comment]
Ok, so I've fixed the part that I believe should be fixed (currently going into the dev branch but will be live with next release). On monster import, bracketed restrictions will be retained in the spell name, because I believe these constitute effectively a different spell to the unrestricted version, and should thus have a different name. I'm still stripping the asterisk, however, since this has nothing to do with the content of the spell itself and shouldn't be part of the name field. I do have this information parsed out, and if you chose to add a field to model this at some point I could populate it easily.
